### ISSUE TITLE
[Bug] no-delegate builtin interpreter corrupts FULLY_CONNECTED output via memory-planner buffer aliasing (XNNPACK/preserve_all_tensors correct)

### ISSUE THREAD

**Issue by pctablet505** [google-ai-edge/LiteRT#8928]
## Summary

The no-delegate builtin TFLite interpreter (`OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES`, the default memory planner) silently returns **wrong values** for a `FULLY_CONNECTED` op's output tensor, even though every one of that op's own inputs is bit-exact. The XNNPACK delegate and `experimental_preserve_all_tensors=True` (which merely disables buffer reuse) both compute the correct result. This is the signature of a memory-planner buffer-aliasing bug: some other tensor's write clobbers this tensor's backing arena buffer before (or instead of) the real computation landing there.

I hit this exporting a multimodal Gemma4 model (PyTorch backend, via `ai_edge_torch`/`litert_torch`) to a `.litertlm` bundle: the vision-encoder branch's final output-projection tensor comes back corrupted under plain interpreter execution, producing wrong KV-cache values (max abs error ~3.0–3.25 across several export runs) at exactly the merged vision-token positions — while the exported graph itself is correct (`torch.export` output matches the original PyTorch/Keras eager model bit-exactly at every position).

## Minimal reproduction

Full standalone repro (a single script — builds its own test model from the TFLite flatbuffer schema, no external binary or data files needed):

**https://gist.github.com/pctablet505/d4313be0187c8074db58b66cf7124835**

The model is an ordinary 6-op graph: `BATCH_MATMUL -> MUL -> RESHAPE -> MUL -> MUL -> FULLY_CONNECTED` — a pure-L2 RMSNorm (mean-of-squares via a matmul-based sum, scale, rsqrt-multiply) followed by a linear/Dense projection. Nothing exotic about the op pattern by itself.

The divergence is fully data-independent — any random input (any seed) at the model's shapes reproduces the same qualitative signature, which points squarely at the memory planner rather than at any special numeric values.

This was found by bisecting (via delta-debugging) a much larger real model export where this exact op sequence appeared as one internal stage of a ~1460-op graph. In that larger graph, roughly 466 unrelated ops executed between producing this stage's RMS-normed tensor and its sole consumer (the `FULLY_CONNECTED`) — i.e. that tensor had to stay alive across many intervening, unrelated ops before being read. That long buffer lifetime, and the resulting competition for arena space, appears necessary to trigger the aliasing: the same 6 ops built in complete isolation, with no other tensors/ops competing for buffer space around them, do not reproduce the bug on their own unless built to retain a similar allocation pattern (which is what the gist's `repro.py` does).

**Expected behavior**: the no-delegate builtin interpreter's output should match XNNPACK / `preserve_all_tensors` output within normal floating-point tolerance (~1e-6).

**Actual behavior**:
```
max_abs_err  no_delegate vs preserve_all: 6.499644e+02
max_abs_err  xnnpack     vs preserve_all: 3.814697e-06
```
The no-delegate output isn't just numerically off — it contains completely different, unrelated-looking values, consistent with reading a stale/aliased buffer from an unrelated computation rather than a precision or kernel-correctness issue.

## Root-cause hypothesis

Memory-planner buffer aliasing: the default planner reuses/aliases the RMS-normed tensor's backing buffer for some other write before the `FULLY_CONNECTED` op (its sole consumer, far later in program order in the original larger graph) reads it. Evidence:
- `experimental_preserve_all_tensors=True` (which disables buffer reuse across ops) fixes it completely.
- The XNNPACK delegate (a completely different execution/buffer-management engine, sharing no buffer-planning code with the builtin interpreter) also computes the correct answer independently.
- Bisecting the original ~1460-op graph by op-count showed the divergence appearing and disappearing depending on exactly which tensor was exposed as the graph's output at a given cut point — i.e. it is highly sensitive to the exact set of tensors alive in the arena at once, not to any op's own semantics.

## Versions

Confirmed on two separate installs:

| | Confirmed-bug env | Latest public env |
|---|---|---|
| `ai_edge_litert` | `2.2.0.dev20260718` | `2.1.6` (newest on PyPI at time of filing; `pip index versions ai-edge-litert` lists no newer build) |
| `numpy` | `2.4.6` | `2.5.1` |
| Python | `3.13.13` | `3.13.13` |
| OS | Ubuntu 26.04 LTS, kernel `7.0.0-28-generic`, x86_64 | same |

The minimal repro above reproduces identically on both versions.

## Additional context

This affects `Interpreter(experimental_op_resolver_type=OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES)` — i.e. any caller who explicitly avoids delegates (e.g. to guarantee CPU-only, delegate-independent execution, or because their model has ops/shapes a delegate can't handle at prepare time) is exposed to this.

Happy to share the full original ~1460-op graph or additional bisection detail if useful for debugging.

**Comment by shivvaam0001** [comment]
Hi @pctablet505,

Thank you for reaching out to us.
I tested this on our end and observed the same behavior. We will investigate it further and get back to you.

Thanks!

**Comment by aravindmurali22** [comment]
Hi @fengwuyao , following up on this thread, is there any updates on this? Thanks!
